content/blog/nats-server-29-release.md

```diff
@@ -0,0 +1,321 @@
+
+This release brings _more_ capabilities to scale in more dimensions and make data locality more transparent.
+
+[subject-messaging]: https://docs.nats.io/nats-concepts/subjects
+
+#### Replica and mirror-based direct gets
+
+Streams have supported a `GetMsg` operation since JetStream originally launched. This method is avaliable on the client's JetStream Manager interface and takes stream name and the specific sequence number to get a _raw_ message from the stream. For example, in Go:
+
+```go
+rmsg, _ := js.GetMsg("EVENTS", 29)
+```
+
+With the introduction of the [Key/Value Store][kv] layer which builds upon a stream and relies on subjects for keys, a new API has been exposed directly in the manager API called `GetLastMsg` which takes the stream name and a _subject_ and returns the last known message for that subject.
+
+```go
+rmsg, _ := js.GetLastMsg("EVENTS", "events.device.15")
+```
+
+With the increasing use cases and potential around the Key/Value API in order to reduce dependencies and infrastructure cost, there has also been a need to further optimize and scale _get_ operations.
+
+Two new **opt-in** stream configuration options include `AllowDirect` and `MirrorDirect`. The first allows get operations to be directed to any replica of the stream, not only the leader, while the second spreads out the gets to any (unfiltered) [mirrors][mirrors] that have been created for that stream as well.
+
+```go
+js.AddStream(&nats.StreamConfig{
+  Name: "EVENTS",
+  Subjects: []string{"events.device.*"},
+  AllowDirect: true,
+  MirrorDirect: true,
+})
+```
+
+Mirrors are of particular interest since they are typically created to provide read-only data locality to clients interested in a stream. For clients using these get operations, requests will be routed to the closest mirror rather than the origin stream.
+
+<img src="/img/nats-server-29-release/direct-get.jpg" width="600px" style="display: block; margin: 0 auto 20px auto" />
+
+Do note that this introduces a trade-off of reducing pressure on the leader to serve both reads and writes exclusively, however at the risk of getting a _stale_ read if the replica or mirror is not fully caught up. A stale read is much less likely among replicas, but for mirrors that there could be increased latency due to being across regions, or if there are network interruptions.
+
+Since mirrors are technically streams themselves with a different name from the origin, each message returned from a *direct get* call will automatically include a few useful headers:
+
+- `Nats-Stream` - original name of the stream the message was written to
+- `Nats-Sequence` - original sequence number in the origin stream
+- `Nats-Time-Stamp` - original timestamp associated with the message
+- `Nats-Subject` - original subject of the message
+- `Nats-Last-Sequence` - the previous sequence number for the subject
```

Comment:
No, this is not part of the direct get response, instead this is in the republish.

Reply:
Thanks for the correction!

Reply:
Thinking about this more, especially with the combination of republish and direct get and reading [this comment](https://github.com/nats-io/nats-server/blob/dc2e4b714a50497ddb42e3075dc57248521da735/server/stream.go#L3880), it seems that the gap detection would only apply for a single message. The pattern of re-connecting a subscription receiving a republishing could use a direct get based on this `Nats-Last-Sequence`, however without additional lookups, the client wouldn't know if yet another message was missed. I suppose at that point, it would be safer for a lightweight consumer to reply and then resume relying on the republish.

Reply:
You would need to store client side all sequences per publish subject. So if you subscribe to foo.*, you would have a map of publish subject to last sequence received.

Reply:
Right, so client knows the last sequence they observed for a given subject, then catch-up accordingly. I was thinking about it in reverse where a republish include the previous sequence, if a direct get on that message would also include the header, then it could daisy chain back until it reaches the sequence.

Reply:
Yes that is correct. A message includes the message sequence and timestamp and the previous sequence for this subject. So you could walk backwards if needed or ask for everything from last sequence you remembered/stored.